source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
+important one is `type`, which specifies what values are valid for an
+option, and how/whether multiple values should be merged together.
+
+In the nixpkgs library there are a number of types available under
+`lib.types`. Here we're using the `lines` type, which specifies that:
+- Only strings are valid values
+- Multiple strings are joined with newlines
+
+```diff
+diff --git a/default.nix b/default.nix
+index f7569bd..cb423a9 100644
+--- a/default.nix
++++ b/default.nix
+@@ -1,3 +1,9 @@
+ { lib, ... }: {
++
++  options = {
++    generate.script = lib.mkOption {
++      type = lib.types.lines;
++    };
++  };
+   
+ }
+```
+
+Let's try to evaluate this with this file:
+
+```nix
+# eval.nix
+(import <nixpkgs/lib>).evalModules {
+  modules = [
+    ./default.nix
+  ];
+}
+```
+
+Then we run
+```bash
+nix-instantiate --eval eval.nix -A config.generate.script
+```
+
+Trying to evaluate the `generate.script` option however, we get an error
+that the option is used but not defined, indicating that we need to
+actually give a value to the option.
+
+
+## Type checking: Assigning integer to generate.script
+
+We can try to assign an integer to our option of type `lines`, but the
+module system correctly throws an error saying that our definition
+doesn't match the options type.
+
+```diff
+diff --git a/default.nix b/default.nix
+index cb423a9..0a9162e 100644
+--- a/default.nix
++++ b/default.nix
+@@ -5,5 +5,9 @@
+       type = lib.types.lines;
+     };
+   };
++
++  config = {
++    generate.script = 42;
++  };
+   
+ }
+```
+
+## Successful evaluation: Assigning a string to generate.script
+
+We can make type checking pass by assigning a string to the option,
+giving us our first successful evaluation of the `generate.script`
+option.
+
+In this case, we assign a script which calls the Google Maps Static API
+to generate a world map, then displaying the result using icat
+(image-cat), both of which are helper scripts.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 0a9162e..24f9c34 100644
+--- a/default.nix
++++ b/default.nix
+@@ -7,7 +7,9 @@
+   };
+ 
+   config = {
+-    generate.script = 42;
++    generate.script = ''
++      map size=640x640 scale=2 | icat
++    '';
+   };
+   
+ }
+```
+
+TODO: Create derivations to get these commands
+
+## A new list option: Declaring generate.requestParams
+
+Let's introduce another option, generate.requestParams. For its type,
+we'll use `listOf <nestedType>`, which is a generic list type where each
+element has to match the given nested type. In our case we want `str` to
+be the nested type, which is a generic string type.
+
+Note that the difference between `str` and `lines` is in their merging
+behavior:
+- For `lines`, multiple definitions get merged by concatenation with
+  newlines
+- For `str`, multiple definitions are not allowed. Which in this case is
+  mostly irrelevant however, since we can't really define a list element
+  multiple times.
```

Comment:
Maybe we should be using _str_ instead of _lines_ at line 70, usually, scripts aren't concatenated anyway.
Then we could move this to line 56
It would reduce complexity (number of concepts) in section of lists.

Reply:
I intentionally tried to introduce as many different types as possible to show what's there. I think it's somewhat common for scripts to be concatenated though, e.g. `systemd.services.<name>.{script,preStart}` is like that

Reply:
> I intentionally tried to introduce as many different types as possible to show what's there.

Agree with this, but we are in a list section, reader have to bash his head against lists, e we're talking about strings,
Ok that computationally speaking a string is a list of chars, and lines is internally represented as a list of strings. But no one needs to know it until the next level. ;-)

Other problem of this, I usually read documentations again trying to find concepts that I know they exist in that document like a reference/dictionary. Include string in list, would be like including the definition of some S word in L page.

That is why suggested move this to line 56 (where we introduce lines)
Also, I was thinking in add other string types strMatching and enum (ok is not a string but mostly used with string)

This suggestion also reduce our str definition that has this _Which in this case is mostly irrelevant however, since we can't really define a list element multiple times._ that may cause confusion, since we are telling something important and saying that it doesn't matter.

Reply:
I think how it's written makes sense. Don't block on this. If anything I would just remove the entire "Which in this case ..." sentence since by it's own admission it's not terribly relevant.